plugins/catalog/src/components/CatalogPage/CatalogPage.tsx

```diff
@@ -46,20 +48,58 @@ const useStyles = makeStyles(theme => ({
 
 import { catalogApiRef } from '../..';
 import { envelopeToComponent } from '../../data/utils';
+import { Component } from '../../data/component';
 
 const CatalogPage: FC<{}> = () => {
   const catalogApi = useApi(catalogApiRef);
   const { value, error, loading } = useAsync(() => catalogApi.getEntities());
-  const [selectedFilter, setSelectedFilter] = React.useState<CatalogFilterItem>(
+  const [locations, setLocations] = useState<any>([]);
+  const [selectedFilter, setSelectedFilter] = useState<CatalogFilterItem>(
     defaultFilter,
   );
 
-  const onFilterSelected = React.useCallback(
+  const onFilterSelected = useCallback(
     selected => setSelectedFilter(selected),
     [],
   );
   const styles = useStyles();
 
+  useEffect(() => {
+    const getLocationDataForEntities = async (entities: Entity[]) => {
+      return Promise.all(
+        entities.map(entity => catalogApi.getLocationByEntity(entity)),
```

Comment:
One afterthought though if we want to continue this road, maybe it's at least worth introducing `batch` endpoint so frontend will at least have it within one roundtrip? @freben @Nek what do you think?

Reply:
I'm more and more leaning towards entities being more self-describing... They could have their actual resolved location added as an annotation. E.g.

`backstage.io/location: github:https://github.com/org/repo/x.yaml`

or similar. Then the refresh loop could look at THAT when refreshing instead. And consumers could trivially locate, link to, process, etc those locations